---
title: "Network Report"
author: "David McGaughey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    theme: flatly
    toc: true
params:
  datafile:
    value: NA
  expression:
    value: NA
  comparison_1:
    value: NA
  comparison_2:
    value: NA

---

Datafile: `r params$datafile`

Differential Expression: `r params$expression`

`r params$comparison_1` against `r params$comparison_2`

Network Analysis
```{r, echo = T, results = 'hide', warning=F}
library(tidyverse, warn.conflicts=F, quietly=T)

#peak <- read_tsv(params$datafile, col_names = F)

#setwd('/Volumes/data/projects/nei/hufnagel/iPSC_RPE_ATAC_Seq/')
known_gfp_ipsc <- read_tsv('homer_unique_peaks_GFP_ATAC-Seq__not__IPSC_ATAC-Seq/all/knownResults_matched_with_RNAseq.tsv')
known_gfp_rfp <- read_tsv('homer_unique_peaks_GFP_ATAC-Seq__not__RFP_ATAC-Seq/all/knownResults_matched_with_RNAseq.tsv')
known_rfp_ipsc <- read_tsv('homer_unique_peaks_RFP_ATAC-Seq__not__IPSC_ATAC-Seq/all/knownResults_matched_with_RNAseq.tsv')

peak_gfp_ipsc <- read_tsv('peak_full/homer_GFP_ATAC-Seq__not__IPSC_ATAC-Seq/all/common_peaks.blackListed.narrowPeak__all_homer_motif.bed.gz', col_names = F) %>% 
  select(X1:X4, X24, X25, X23, X26:X31)
colnames(peak_gfp_ipsc) <- c('chrom','start','stop','peak_score', 'Gene', 'Description', 'Distance_TSS_peak', 'chrom_motif','start_motif','stop_motif','motif','motif_score','strand_motif')
peak_gfp_rfp <- read_tsv('peak_full/homer_GFP_ATAC-Seq__not__RFP_ATAC-Seq/all/common_peaks.blackListed.narrowPeak__all_homer_motif.bed.gz', col_names = F) %>% 
  select(X1:X4, X24, X25, X23, X26:X31)
colnames(peak_gfp_rfp) <- c('chrom','start','stop','peak_score', 'Gene', 'Description', 'Distance_TSS_peak', 'chrom_motif','start_motif','stop_motif','motif','motif_score','strand_motif')
peak_rfp_ipsc <- read_tsv('peak_full/homer_RFP_ATAC-Seq__not__IPSC_ATAC-Seq/all/common_peaks.blackListed.narrowPeak__all_homer_motif.bed.gz', col_names = F) %>% 
  select(X1:X4, X24, X25, X23, X26:X31)
colnames(peak_rfp_ipsc) <- c('chrom','start','stop','peak_score', 'Gene', 'Description', 'Distance_TSS_peak', 'chrom_motif','start_motif','stop_motif','motif','motif_score','strand_motif')

# expression %>% filter(Gene %in% (known_rfp_ipsc %>% filter(`log P-pvalue` < -1000) %>% filter(Name %in% (known_gfp_ipsc %>% filter(`log P-pvalue` < -1000) %>% pull(Name))) %>% pull(Gene))) %>% data.frame() %>% arrange(Gene, Line)

expression_gfp_ipsc <- read_csv('~/git/ipsc_rpe_RNA-seq/data/GFP_vs_iPSC.results.csv')
expression_gfp_rfp <- read_csv('~/git/ipsc_rpe_RNA-seq/data/GFP_vs_RFP.results.csv')
expression_rfp_ipsc <- read_csv('~/git/ipsc_rpe_RNA-seq/data/RFP_vs_iPSC.results.csv')
expression_rpe_ipsc <- read_csv('~/git/ipsc_rpe_RNA-seq/data/RPE_vs_iPSC.results.csv')

expression <- read_tsv('~/git/ipsc_rpe_RNA-seq/data/lsTPM_by_Line.tsv')

top_motifs_rpe_ipsc <- known_rfp_ipsc %>% filter(`log P-pvalue` < -1000) %>% filter(Name %in% (known_gfp_ipsc %>% filter(`log P-pvalue` < -1000) %>% pull(Name)))
top_motifs_gfp_rfp <- known_gfp_rfp %>% filter(`log P-pvalue` < -100) 

```

# Top TFBS that are highly enriched in both GFP and RFP compared to iPSC
```{r}
top_motifs_rpe_ipsc %>% mutate(Gene = case_when(grepl('^FRA', Gene) ~ gsub('FRA', 'FOSL', Gene),
                                                grepl('AP1', Gene) ~ 'JUN',
                                                TRUE ~ Gene)) %>%  
  select(Rank:RFP) %>% left_join(expression_rpe_ipsc) %>% DT::datatable()
```

# Top TFBS that are highly enriched in both GFP compared to RFP
```{r}
top_motifs_gfp_rfp %>% select(Rank:RFP) %>% left_join(expression_gfp_rfp) %>% DT::datatable()
```

```{r}
# match TF to genes
colnames(peak_gfp_ipsc) <- c('chrom','start','stop','size','macs2_score','strand','thickStart','thickStop','RGB','Distance_to_Gene','Gene','Description','space', 'chrom_motif','start_motif','stop_motif','motif','motif_score','strand_motif')

# create gene centric info on peak/motif and differential gene expression iPSC <-> RPE
grouper <- function(peak, motifs){
  grouped_stats <- list()
  for (i in motifs){
    out <- peak %>% 
      filter(peak_score > 75) %>% 
      filter(!is.na(Gene)) %>% # remove NA genes where the peak didn't get a match (no genes within 500kb of a peak)
      filter(motif == i) %>% # only retain the motif TFBS homer hits
      group_by(chrom, start, stop, Gene) %>% # group by GFP/RFP,iPSC, the macs2 peak, and the gene
      summarise(Num_TFBS_motifs=n()) %>%  # count total number of motifs
      group_by(Gene) %>% # group now by GFP/RFP/IPSC and gene
      summarise(Total_TFBS_motifs = sum(Num_TFBS_motifs), Num_Peaks = n()) %>% # now add num of peaks near gene
      select(Gene, Total_TFBS_motifs, Num_Peaks)
    grouped_stats[[i]] <- out
    print(i)
  }
  grouped_stats
}

gfp_ipsc_list <- grouper(peak_gfp_ipsc, top_motifs_gfp_rfp$Name)
rfp_ipsc_list <- grouper(peak_rfp_ipsc, top_motifs_gfp_rfp$Name)

# cut down to diff expressed genes and merge gfp_ipsc_list and rfp_ipsc_list
rpe_ipsc_list <- list()
for (i in names(gfp_ipsc_list)){
  print(i)
  rpe_ipsc_list[[i]] <- rfp_ipsc_list[[i]] %>% 
    left_join(expression_rfp_ipsc) %>% 
    filter(padj < 0.01) %>% 
    filter(Gene %in% (gfp_ipsc_list[[i]] %>% 
                        left_join(expression_gfp_ipsc) %>% 
                        filter(padj < 0.01) %>% pull(Gene)))
}

# identify TFBS with overlapping gene lists
overlap_df <- data.frame(i = character(), y = character(), ratio = double())
overlap_matrix <- matrix(ncol = length(names(rpe_ipsc_list)), nrow = length(names(rpe_ipsc_list)))
row.names(overlap_matrix) <- names(rpe_ipsc_list)
colnames(overlap_matrix) <- names(rpe_ipsc_list)
for (i in names(rpe_ipsc_list)){
  row <- c()
  for (j in names(rpe_ipsc_list)){
    ratio <-  rpe_ipsc_list[[i]]$Gene %in% rpe_ipsc_list[[j]]$Gene %>% sum() / 
      nrow(rpe_ipsc_list[[i]])
    row <- c(row, ratio)
  }
  overlap_matrix[i,] <- row
}

library(ComplexHeatmap)
Heatmap(overlap_matrix, column_names_max_height = unit(20, 'cm'),  row_names_max_width = unit(20, 'cm'))

TF_gene_stats <- bind_rows(rpe_ipsc_list, .id = 'motif') %>% rowwise() %>% mutate(motif_core = toupper(gsub('\\(.*|.*BestGuess:','',motif))) %>% ungroup()
```

TF network
```{r}
library(visNetwork, warn.conflicts=F, quietly=T)



way_too_common_genes <- TF_gene_stats %>% group_by(Gene) %>% summarise(Count= n()) %>% 
  filter(Count > 15) %>% pull(Gene)




TF_gene_stats_n <- TF_gene_stats %>% 
  left_join(expression %>% select(Gene, TF)) %>% filter(TF == 'Yes') %>% unique() %>% 
  filter(!Gene  %in% way_too_common_genes)



TF <- TF_gene_stats_n %>%
  distinct(motif_core) %>%
  rename(label = motif_core)

genes <- TF_gene_stats_n %>%
  
  distinct(Gene) %>%
  rename(label = Gene)

expression_TF <- left_join(expression, tf_gene, by = 'Gene')

nodes <- full_join(TF, genes, by = "label") %>%  
  arrange(label) %>% rowid_to_column("id") 


edges <- TF_gene_stats_n %>%  
  select(TF = motif_core, Gene, weight = log2FoldChange) %>% 
  unique() %>% 
  left_join(nodes, by = c("TF" = "label")) %>% 
  rename(from = id) %>% 
  left_join(nodes, by = c("Gene" = "label")) %>% 
  rename(to = id) %>% 
  ungroup() %>% 
  select(from, to, weight) 


















nodes <- TF_gene_stats %>% left_join(expression %>% select(Gene, TF)) %>% filter(TF == 'Yes') %>% select(label = Gene) %>% unique()






edges <- TF_gene_stats %>%  
  select(TF = motif_core, Gene, weight = log2FoldChange) %>% 
  unique() %>% 
  left_join(nodes, by = c("TF" = "label")) %>% 
  rename(from = TF, to = Gene)

%>% 
  left_join(nodes, by = c("Gene" = "label")) %>% 
  rename(to = id) %>% 
  ungroup() %>% 
  select(from, to, weight) 

visNetwork(nodes, edges, height = "700px", width = "100%") %>% 
  visGroups(groupname = 'TF', color='red') %>% 
  visGroups(groupname = 'goTF', color = 'pink') %>% 
  visOptions(highlightNearest = list(enabled = T, degree = 2),
             nodesIdSelection = TRUE) %>% 
  visPhysics(stabilization = TRUE) %>% 
  visEdges(smooth = FALSE) %>% 
  visInteraction(navigationButtons = TRUE)

1
```

# Full table
The GFP/iPSC/RFP columns are the number of peaks with the motif associated (by distance) to the gene

Ratio is GFP/(RFP or iPSC)

Delta is GFP - (RFP or iPSC)

If log2FC is positive, then the expression of the gene is higher in GFP vs (RFP or iPSC)
```{r}
TF_gene_stats %>% dplyr::select(motif = motif_core, Gene:padj, motif_full = motif) %>% DT::datatable( filter = list(position = 'top', clear = FALSE),
                                                                                                      extensions = 'Buttons', options = list(
                                                                                                        lengthMenu = list(c(5, 15, 30, 60, 100, -1), list('5', '15', '30', '60', '100', 'All')),
                                                                                                        dom = 'Bfrtip',
                                                                                                        buttons = c('copy', 'csv', 'excel', 'pageLength')
                                                                                                      ))
```

# Count number of genes associated with each TFBS
```{r}
TF_gene_stats %>% group_by(motif_core) %>% summarise(count = n(), Gene = paste(Gene, collapse = ',') ) %>% arrange(-count) %>% DT::datatable()
```

# Most common genes associated with TF
```{r}
TF_gene_stats %>% group_by(Gene) %>% summarise(count = n(), motif = paste(motif_core, collapse=',')) %>% arrange(-count) %>%  DT::datatable( filter = list(position = 'top', clear = FALSE))
```


# Network
No more than 5 genes assigned to a single TF. Picking the top 5 genes by delta of `r params$comparison_1` peaks - `r params$comparison_2` peaks, then breaking ties by taking the most differentially expressed by bulk RNA-seq RPE vs iPSC. 

Red are the homer ID'ed TF(BS)

Pink are genes labelled as TF by GO term (but not ID'ed by homer)

If you hover you get the log2FC

You can zoom in and out

Clicking on a node gives you all the other nodes within 2 edges
```{r, fig.height=6, fig.width=6}
library(visNetwork, warn.conflicts=F, quietly=T)
TF_gene_stats_n <- TF_gene_stats %>% 
  # filter(!grepl('GATA', motif_core)) %>% 
  group_by(motif_core) %>% 
  filter(delta > 2) %>% 
  # top_n(5, delta) %>% # no more than n by tF 
  mutate(`(abs)log2FC` = abs(log2FoldChange)) %>% 
  #top_n(5,`(abs)log2FC`) %>% 
  #filter(!Gene %in% c('FZD8','OPCML','PFKP','NRIP1','TSHZ3', 'SOX6')) %>% # trimming out most common genes between TF
  ungroup()

TF <- TF_gene_stats_n %>%
  distinct(motif_core) %>%
  rename(label = motif_core)

genes <- TF_gene_stats_n %>%
  distinct(Gene) %>%
  rename(label = Gene)

expression_TF <- left_join(expression, tf_gene, by = 'Gene')

nodes <- full_join(TF, genes, by = "label") %>%  
  arrange(label) %>% rowid_to_column("id") %>% 
  mutate(group = case_when(label %in% TF$label ~ 'TF', 
                           TRUE ~ 'Gene')) %>% 
  left_join(., expression_TF %>% 
              select(Gene, value = log2FoldChange, goTF = TF) %>% 
              unique(), by = c("label" = "Gene")) %>% 
  mutate(title = paste0('Gene: ', label, '<br>log2FC: ', round(value,2))) %>% 
  mutate(value = 10 ** value) %>% 
  mutate(group = case_when(group == 'TF' ~ 'TF',
                           goTF == 'Yes' ~ 'goTF',
                           TRUE ~ 'Gene'))


edges <- TF_gene_stats_n %>%  
  select(TF = motif_core, Gene, weight = log2FoldChange) %>% 
  unique() %>% 
  left_join(nodes, by = c("TF" = "label")) %>% 
  rename(from = id) %>% 
  left_join(nodes, by = c("Gene" = "label")) %>% 
  rename(to = id) %>% 
  ungroup() %>% 
  select(from, to, weight) 

visNetwork(nodes, edges, height = "700px", width = "100%") %>% 
  visGroups(groupname = 'TF', color='red') %>% 
  visGroups(groupname = 'goTF', color = 'pink') %>% 
  visOptions(highlightNearest = list(enabled = T, degree = 2),
             nodesIdSelection = TRUE) %>% 
  visPhysics(stabilization = TRUE) %>% 
  visEdges(smooth = FALSE) %>% 
  visInteraction(navigationButtons = TRUE)
```

# GO Term Enrichment of Gene List
```{r, echo = T, warning=F, message = F}
ids <- clusterProfiler::bitr(expression$Gene, fromType="SYMBOL", toType=c("UNIPROT", "ENSEMBL", "ENTREZID"), OrgDb="org.Hs.eg.db")

set <- clusterProfiler::bitr(TF_gene_stats$Gene %>% unique(),
                             fromType="SYMBOL",
                             toType="ENTREZID",
                             OrgDb = "org.Hs.eg.db")




ego_bp <- clusterProfiler::enrichGO(gene          = set$ENTREZID,
                                    universe      = ids$ENTREZID %>% unique(),
                                    OrgDb         = org.Hs.eg.db,
                                    ont           = "BP",
                                    pool = TRUE,
                                    pAdjustMethod = "BH",
                                    pvalueCutoff = 0.1,
                                    qvalueCutoff = 0.1,
                                    keyType = 'ENTREZID',
                                    readable      = TRUE)

ego2_bp <- clusterProfiler::simplify(ego_bp, cutoff=0.7, by="p.adjust", select_fun=min)

ego_mf <- clusterProfiler::enrichGO(gene          = set$ENTREZID,
                                    universe      = ids$ENTREZID %>% unique(),
                                    OrgDb         = org.Hs.eg.db,
                                    ont           = "MF",
                                    pool = TRUE,
                                    pAdjustMethod = "BH",
                                    qvalueCutoff = 0.1,
                                    keyType = 'ENTREZID',
                                    readable      = TRUE)

ego2_mf <- clusterProfiler::simplify(ego_mf, cutoff=0.7, by="p.adjust", select_fun=min)

ego2_bp_out <- ego2_bp %>% data.frame() %>%  mutate(geneID = gsub('/', '<br>', geneID)) %>%  mutate_at(.vars=vars(pvalue,p.adjust,qvalue), .funs=funs(formatC(., format = "e", digits = 2))) 

ego2_mf_out <- ego2_mf %>% data.frame() %>%  mutate(geneID = gsub('/', '<br>', geneID)) %>%  mutate_at(.vars=vars(pvalue,p.adjust,qvalue), .funs=funs(formatC(., format = "e", digits = 2)))

bind_rows(ego2_bp_out %>% mutate(ONT = 'BP'), ego2_mf_out %>% mutate(ONT = 'MF')) %>% arrange(pvalue) %>% DT::datatable(rownames = FALSE, escape = FALSE)
```

# KEGG Pathway Enrichment
```{r, fig.width=5}
kegg <- clusterProfiler::enrichKEGG(gene         = set$ENTREZID,
                                    organism     = 'hsa', 
                                    pvalueCutoff = 1)
kegg <- DOSE::setReadable(kegg,  OrgDb = org.Hs.eg.db, keytype = 'ENTREZID')

kegg %>% data.frame() %>%  
  mutate(geneID = gsub('/', '<br>', geneID)) %>%  
  mutate_at(.vars=vars(pvalue,p.adjust,qvalue), .funs=funs(formatC(., format = "e", digits = 2))) %>% 
  DT::datatable(rownames = FALSE, escape = FALSE) 

kegg %>% data.frame() %>% mutate(Terms = paste(ID, Description)) %>% head(10) %>% mutate(Terms =factor(Terms, levels = Terms %>% rev())) %>%  ggplot(aes(x=Terms, y= -log10(pvalue))) + geom_bar(stat='identity') + theme_minimal() + coord_flip() + xlab('-log10(p value')

