---
title: "Pan Condition Trends"
author: "David McGaughey"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_notebook:
    theme: flatly
    toc: true
---

Compare genes with expression changes (bulk RNA-seq)
  iPSC -> RFP -> GFP

correlate with “chromatin openness/closingness” (peak changes)

Identify genes where the chromatin changes are consistent iPSC -> RFP -> GFP

Then in those differential peak areas, see what TFBS are in them

4 groups of TFs tagged to genes changing expression/chromatin changes

  i.      Gene expression decreasing, #TFBS chromatin opening
  
  ii.      Gene expression decreasing, #TFBS chromatin closing
  
  iii.      Gene expression increasing, #TFBS chromatin opening
  
  iv.      Gene expression increasing, #TFBS chromatin closing
  
Idea is to identify good TF targets for lentiviral knockdown or overexpression to turn on/off GFP

```{r, warning = F, message = F}
library(tidyverse)
library(ComplexHeatmap)
library(viridis)
library(circlize)
library(annotables)
#peak <- read_tsv('/Volumes/data/projects/nei/hufnagel/iPSC_RPE_ATAC_Seq/peak_full/homer_GFP__not__IPSC/all/all_common_peaks.blackListed.narrowPeak.closestTSS__interesting_homer_motif.bed.gz')
expression <- read_csv('~/git/ipsc_rpe_RNA-seq/data/lsTPM_by_Line_with_Diff_Exp.tsv')
cgm_gfp <- read_tsv('/Volumes/data/projects/nei/hufnagel/iPSC_RPE_ATAC_Seq/CGM/GFP_common_peaks.blackListed.narrowPeak.CGM.tsv') %>% 
  left_join(. , expression %>% filter(Line == 'GFP') %>% select(Gene, lsTPM), by = c('Gene_Name' = 'Gene'))
cgm_ipsc <- read_tsv('/Volumes/data/projects/nei/hufnagel/iPSC_RPE_ATAC_Seq/CGM/IPSC_common_peaks.blackListed.narrowPeak.CGM.tsv') %>% 
  left_join(. , expression %>% filter(Line == 'iPSC') %>% select(Gene, lsTPM), by = c('Gene_Name' = 'Gene'))
cgm_rfp <-  read_tsv('/Volumes/data/projects/nei/hufnagel/iPSC_RPE_ATAC_Seq/CGM/RFP_common_peaks.blackListed.narrowPeak.CGM.tsv')  %>% 
  left_join(. , expression %>% filter(Line == 'RFP') %>% select(Gene, lsTPM), by = c('Gene_Name' = 'Gene'))


delta_maker <- function(df1, df2){
  delta <- (df1 %>% mutate(`log2(TPM+1)` = log2(lsTPM+1)) %>%  
              select(Exon_Count:Gene_length, `log2(TPM+1)`)) - 
    (df2 %>% mutate(`log2(TPM+1)` = log2(lsTPM+1)) %>% 
       select(Exon_Count:Gene_length, `log2(TPM+1)`)) %>% 
    as.tibble()
  delta$Gene_Name = df1$Gene_Name
  
  delta <- delta %>% mutate(Promoter = `5000` + `1000`,
                            Upstream100K = `1000` + `5000` + `1e4` + `5e4` + `1e5`)
  
  delta %>% select(Exon_Count:`-1e6`) %>% rowSums(na.rm=TRUE) -> delta$All
  delta
}

rfp_ipsc <- delta_maker(cgm_rfp, cgm_ipsc)
gfp_rfp <- delta_maker(cgm_gfp, cgm_rfp)
```

```{r}
heatmap_maker <- function(df1, df2, union_names){

  pattern1 <- df1 %>% 
    filter(Gene_Name %in% union_names) %>% 
    select(Gene_Name, Promoter, Upstream100K, All, `Delta Expression` = `log2(TPM+1)`) %>% 
    left_join(grch38 %>% select(symbol, description), by = c("Gene_Name" = "symbol")) %>%
    unique()
  row.names(pattern1) = pattern1$Gene_Name
  
  pattern2 <- df2 %>% 
    filter(Gene_Name %in% union_names) %>% 
    select(Gene_Name, Promoter, Upstream100K, All, `Delta Expression` = `log2(TPM+1)`) %>% 
    left_join(grch38 %>% select(symbol, description), by = c("Gene_Name" = "symbol")) %>% 
    unique()
  row.names(pattern2) = pattern2$Gene_Name
  
  
  one <- pattern1 %>% select(Promoter:All) %>% 
    Heatmap(col = colorRamp2(c(-2, 0, 2), viridis(3)), 
            cluster_columns = F,
            cluster_rows = F,
            show_row_names = F, 
            column_title = 'GFP - RFP',
            name = 'Delta ATAC Peak')
  two <-  pattern1 %>% select(`Delta Expression`) %>% 
    Heatmap(col = colorRamp2(c(-2, 0, 2), magma(3)), 
            cluster_columns = F,
            name = 'Delta log2(TPM+1)',
            cluster_rows = F)
  
  three <- pattern2 %>% select(Promoter:All) %>% 
    Heatmap(col = colorRamp2(c(-2, 0, 2), viridis(3)), 
            cluster_columns = F,
            cluster_rows = F,
            show_row_names = F,
            show_heatmap_legend = F,
            column_title = 'RFP - iPSC')
  
  four <-  pattern2 %>% select(`Delta Expression`) %>% 
    Heatmap(col = colorRamp2(c(-2, 0, 2), magma(3)), 
            cluster_columns = F,
            cluster_rows = F,
            show_heatmap_legend = F,
            show_row_names = F)
  
  three + four + one + two
}
```
More peaks, more expression iPSC -> RFP -> GFP
```{r}
  
names <- rfp_ipsc %>% filter(`log2(TPM+1)` > 0.5) %>% filter(Promoter > 1 | Upstream100K > 1 | All > 1) %>% pull(Gene_Name)

union_names <- gfp_rfp %>% filter(Gene_Name %in% names, `log2(TPM+1)` > 0.5) %>% filter(Promoter > 1 | Upstream100K > 1 | All > 1) %>% pull(Gene_Name)


heatmap_maker(gfp_rfp, rfp_ipsc, union_names)
```

More peaks, less expression iPSC -> RFP -> GFP
```{r}
names <- rfp_ipsc %>% filter(`log2(TPM+1)` < -0.5) %>% filter(Promoter > 1 | Upstream100K > 1 | All > 1) %>% pull(Gene_Name)

union_names <- gfp_rfp %>% filter(Gene_Name %in% names, `log2(TPM+1)` < -0.5) %>% filter(Promoter > 1 | Upstream100K > 1 | All > 1) %>% pull(Gene_Name)

heatmap_maker(gfp_rfp, rfp_ipsc, union_names)
```

Fewer peaks, more expression iPSC -> RFP -> GFP
```{r, fig.height=10, fig.width=3}
names <- rfp_ipsc %>% filter(`log2(TPM+1)` > 0.5) %>% filter(Promoter < -1 | Upstream100K < -1 | All < -1) %>% pull(Gene_Name)

union_names <- gfp_rfp %>% filter(Gene_Name %in% names, `log2(TPM+1)` > 0.5) %>% filter(Promoter < -1 | Upstream100K < -1 | All < -1) %>% pull(Gene_Name) %>% unique()

heatmap_maker(gfp_rfp, rfp_ipsc, union_names)
```

Fewer peaks, less expression iPSC -> RFP -> GFP
```{r, fig.height=12, fig.width=3}
names <- rfp_ipsc %>% filter(`log2(TPM+1)` < -0.5) %>% filter(Promoter < -1 | Upstream100K < -1 | All < -1) %>% pull(Gene_Name)

union_names <- gfp_rfp %>% filter(Gene_Name %in% names, `log2(TPM+1)` < -0.5) %>% filter(Promoter < -1 | Upstream100K < -1 | All < -1) %>% pull(Gene_Name) %>% unique()
heatmap_maker(gfp_rfp, rfp_ipsc, union_names)
```